engines/wintermute/base/gfx/x/animation.cpp

```diff
@@ -0,0 +1,393 @@
+				keyIndex2 = key;
+				if (key > 0) {
+					keyIndex1 = key - 1;
+				} else { // when ikey == 0, then dwp2 == 0
+					keyIndex1 = key;
+				}
+
+				break;
+			}
+		}
+		time1 = _posKeys[keyIndex1]->_time;
+		time2 = _posKeys[keyIndex2]->_time;
+
+		// get the lerp value
+		if (time2 - time1 == 0)
+			lerpValue = 0;
+		else
+			lerpValue = float(localTime - time1) / float(time2 - time1);
+
+		resultPos = (1 - lerpValue) * _posKeys[keyIndex1]->_pos + lerpValue * _posKeys[keyIndex2]->_pos;
+
+		animate = true;
+	}
+
+	if (animate) {
+		_targetFrame->setTransformation(slot, resultPos, resultScale, resultRot, animLerpValue);
+	}
+
+	return true;
+}
+
+//////////////////////////////////////////////////////////////////////////
+int Animation::getFrameTime() {
+	uint32 frameTime = 0;
+	uint32 prevTime;
+
+	// get the shortest frame time
+	prevTime = 0;
+	for (uint32 i = 0; i < _rotKeys.size(); i++) {
+		if (frameTime == 0 || _rotKeys[i]->_time - prevTime < frameTime)
+			frameTime = _rotKeys[i]->_time - prevTime;
+
+		prevTime = _rotKeys[i]->_time;
+	}
+
+	prevTime = 0;
+	for (uint32 i = 0; i < _posKeys.size(); i++) {
+		if (frameTime == 0 || _posKeys[i]->_time - prevTime < frameTime)
+			frameTime = _posKeys[i]->_time - prevTime;
+
+		prevTime = _posKeys[i]->_time;
+	}
+
+	prevTime = 0;
+	for (uint32 i = 0; i < _scaleKeys.size(); i++) {
+		if (frameTime == 0 || _scaleKeys[i]->_time - prevTime < frameTime)
+			frameTime = _scaleKeys[i]->_time - prevTime;
+
+		prevTime = _scaleKeys[i]->_time;
+	}
+
+	return frameTime;
+}
+
+//////////////////////////////////////////////////////////////////////////
+uint32 Animation::getTotalTime() {
+	uint32 totalTime = 0;
+	if (_rotKeys.size() > 0) {
+		totalTime = MAX(totalTime, _rotKeys[_rotKeys.size() - 1]->_time);
+	}
+
+	if (_posKeys.size() > 0) {
+		totalTime = MAX(totalTime, _posKeys[_posKeys.size() - 1]->_time);
+	}
+
+	if (_scaleKeys.size() > 0) {
+		totalTime = MAX(totalTime, _scaleKeys[_scaleKeys.size() - 1]->_time);
+	}
+
+	return totalTime;
+}
+
+// try to put these functions into a template?
+bool Animation::loadRotationKeyData(XFileLexer &lexer, int count) {
+	for (int keyIndex = 0; keyIndex < count; ++keyIndex) {
+		BoneRotationKey *key = new BoneRotationKey;
+		key->_time = readInt(lexer);
+
+		// should always be four
+		int floatCount = readInt(lexer);
+
+		// .X file format puts the w coordinate first
+		key->_rotation.w() = readFloat(lexer);
+		key->_rotation.x() = readFloat(lexer);
+		key->_rotation.y() = readFloat(lexer);
+		// mirror z component
+		key->_rotation.z() = -readFloat(lexer);
+
+		lexer.advanceToNextToken(); // skip semicolon
+		lexer.advanceToNextToken(); // skip closed braces
+
+		_rotKeys.push_back(key);
+	}
+
+	return true;
+}
+
+bool Animation::loadScaleKeyData(XFileLexer &lexer, int count) {
+	for (int keyIndex = 0; keyIndex < count; ++keyIndex) {
+		BoneScaleKey *key = new BoneScaleKey;
+		key->_time = readInt(lexer);
+
+		// should always be three

```

Comment:
Should we add an assertion about that?

Reply:
this should be included in this PR, please add assert

Reply:
Done